File: apps/cxx11/example_io_app/example_io_app.cxx
/*
* (c) Copyright, Real-Time Innovations, 2025.  All rights reserved.
* RTI grants Licensee a license to use, modify, compile, and create derivative
* works of the software solely for use with RTI Connext DDS. Licensee may
* redistribute copies of the software provided that all such copies are subject
* to this license. The software is provided "as is", with no warranty of any
* type, including any warranty for fitness for any purpose. RTI is under no
* obligation to maintain or support the software. RTI shall not be liable for
* any incidental or consequential damages arising out of the use or inability
* to use the software.
*/

#include <iostream>
#include <sstream>
#include <thread>
#include <chrono>
#include <atomic>
#include <ctime>

// include both the standard APIs and extensions
#include <rti/rti.hpp>
#include <rti/core/cond/AsyncWaitSet.hpp>
#include <rti/distlogger/DistLogger.hpp>
#include <rti/config/Logger.hpp>

//
// For more information about the headers and namespaces, see:
//    https://community.rti.com/static/documentation/connext-dds/7.3.0/doc/api/connext_dds/api_cpp2/group__DDSNamespaceModule.html
// For information on how to use extensions, see:
//    https://community.rti.com/static/documentation/connext-dds/7.3.0/doc/api/connext_dds/api_cpp2/group__DDSCpp2Conventions.html

#include "application.hpp"  // for command line parsing and ctrl-c
#include "ExampleTypes.hpp"
#include "Definitions.hpp"
#include "DDSParticipantSetup.hpp"
#include "DDSReaderSetup.hpp"
#include "DDSWriterSetup.hpp"

// For example legibility.
using namespace rti::all;
using namespace rti::dist_logger;

constexpr int ASYNC_WAITSET_THREADPOOL_SIZE = 5;
const std::string APP_NAME = "Example CXX IO APP";


void process_command_data(dds::sub::DataReader<example_types::Command> reader)
{
    auto samples = reader.take();
    for (const auto& sample : samples)
    {
      // Check if message is not DDS metadata
      if (sample.info().valid())
      {

        // Do something with data message (logged at DEBUG level)
        std::ostringstream oss;
        oss << sample.data();
        rti::config::Logger::instance().debug(("[COMMAND] " + oss.str()).c_str());

        // overloaded -> operator to use RTI extension
        rti::config::Logger::instance().debug(("[COMMAND] Topic '" + std::string(reader->topic_name()) + "' received").c_str());
      }
    }
}

void on_command_liveliness_changed(dds::sub::DataReader<example_types::Command> reader)
{
    auto status = reader->liveliness_changed_status();
    rti::config::Logger::instance().notice(
        ("[COMMAND] Liveliness changed - alive_count: " +
         std::to_string(status.alive_count()) + ", not_alive_count: " +
         std::to_string(status.not_alive_count())).c_str());
}


void process_button_data(dds::sub::DataReader<example_types::Button> reader)
{
    auto samples = reader.take();
    for (const auto& sample : samples)
    {
      // Check if message is not DDS metadata
      if (sample.info().valid())
      {

        // Do something with data message
        std::ostringstream oss;
        oss << sample.data();
        rti::config::Logger::instance().debug(("[BUTTON] " + oss.str()).c_str());

        // overloaded -> operator to use RTI extension
        rti::config::Logger::instance().debug(("[BUTTON] Topic '" + std::string(reader->topic_name()) + "' received").c_str());
      }
    }
} 

void process_config_data(dds::sub::DataReader<example_types::Config> reader)
{
    auto samples = reader.take();
    for (const auto& sample : samples)
    {
      // Check if message is not DDS metadata
      if (sample.info().valid())
      {

        // Do something with data message
        std::ostringstream oss;
        oss << sample.data();
        rti::config::Logger::instance().debug(("[CONFIG] " + oss.str()).c_str());

        // overloaded -> operator to use RTI extension
        rti::config::Logger::instance().debug(("[CONFIG] Topic '" + std::string(reader->topic_name()) + "' received").c_str());
      }
    }
}

void on_position_publication_matched(dds::pub::DataWriter<example_types::Position> writer)
{
    auto status = writer->publication_matched_status();
    rti::config::Logger::instance().notice(
        ("[POSITION] Publication matched - current_count: " +
         std::to_string(status.current_count()) + ", total_count: " +
         std::to_string(status.total_count())).c_str());
}


void run(std::shared_ptr<DDSParticipantSetup> participant_setup)
{
    rti::config::Logger::instance().notice(("Example I/O application starting on domain " + std::to_string(participant_setup->participant().domain_id())).c_str());

    // DDSReaderSetup and DDSWriterSetup are example wrapper classes for your convenience that simplify
    // DDS reader/writer creation and event handling. They manage DataReader/DataWriter lifecycle, attach
    // status conditions to the centralized AsyncWaitSet, and provide convenient methods to register
    // callbacks for DDS events (data_available, subscription_matched, liveliness_changed, etc.)

    // Setup Reader Interfaces
    auto command_reader = std::make_shared<DDSReaderSetup<example_types::Command>>(
        participant_setup,
        topics::COMMAND_TOPIC,
        qos_profiles::ASSIGNER);

    auto button_reader = std::make_shared<DDSReaderSetup<example_types::Button>>(
        participant_setup,
        topics::BUTTON_TOPIC,
        qos_profiles::ASSIGNER);

    auto config_reader = std::make_shared<DDSReaderSetup<example_types::Config>>(
        participant_setup,
        topics::CONFIG_TOPIC,
        qos_profiles::ASSIGNER);

    // Setup Writer Interfaces
    auto position_writer = std::make_shared<DDSWriterSetup<example_types::Position>>(
        participant_setup,
        topics::POSITION_TOPIC,
        qos_profiles::ASSIGNER);

    // Enable Asynchronous Event-Driven processing for readers
    command_reader->set_data_available_handler(process_command_data);
    command_reader->set_liveliness_changed_handler(on_command_liveliness_changed);
    button_reader->set_data_available_handler(process_button_data);
    config_reader->set_data_available_handler(process_config_data);

    // Set publication matched callback for writer
    position_writer->set_publication_matched_handler(on_position_publication_matched);

    rti::config::Logger::instance().notice("Example I/O app is running. Press Ctrl+C to stop.");
    rti::config::Logger::instance().notice("Subscribing to Command, Button, and Config messages...");
    rti::config::Logger::instance().notice("Publishing Position messages...");


    example_types::Position pos_msg;
    pos_msg.source_id(APP_NAME);

    // Counter for tracking iterations
    int iteration = 0;

    while (!application::shutdown_requested) {

      try
      {
        // Populate and send position message
        pos_msg.latitude(37.7749);
        pos_msg.longitude(-122.4194);
        pos_msg.altitude(15.0);
        pos_msg.timestamp_sec(static_cast<int32_t>(std::time(nullptr)));
        position_writer->writer().write(pos_msg);

        // Log every position publish at DEBUG level (can be filtered)
        rti::config::Logger::instance().debug(("[POSITION] Published ID: " + std::string(pos_msg.source_id()) +
                    ", Lat: " + std::to_string(pos_msg.latitude()) +
                    ", Lon: " + std::to_string(pos_msg.longitude()) +
                    ", Alt: " + std::to_string(pos_msg.altitude()) + "m" +
                    ", Timestamp: " + std::to_string(pos_msg.timestamp_sec())).c_str());
        
        // Every 10 iterations (5 seconds), log INFORMATIONAL level status to distributed logger
        if (iteration % 10 == 0) {
          rti::config::Logger::instance().informational(("Application running - Position published at " + 
                      std::to_string(pos_msg.timestamp_sec())).c_str());
        }
        
        iteration++;
      }
      catch (const std::exception &ex)
      {
        rti::config::Logger::instance().error(("Failed to publish position: " + std::string(ex.what())).c_str());
      }

      // Alternate Option: Use Polling Method to Read Data
      // Latency contingent on loop rate
      // process_command_data(command_reader->reader());

      // Sleep
      std::this_thread::sleep_for(std::chrono::milliseconds(500));

    }

    rti::config::Logger::instance().informational("Example I/O application shutting down...");
    rti::config::Logger::instance().notice("Example I/O application stopped");
}

int main(int argc, char *argv[])
{
    using namespace application;

    // Parse arguments and handle control-C
    auto arguments = parse_arguments(argc, argv);
    if (arguments.parse_result == ParseReturn::exit) {
        return EXIT_SUCCESS;
    } else if (arguments.parse_result == ParseReturn::failure) {
        return EXIT_FAILURE;
    }
    setup_signal_handlers();

    try {
        // Create DDS Participant Setup (creates DomainParticipant and AsyncWaitSet)
        // DDSParticipantSetup is an example wrapper class for your convenience that manages the DDS
        // infrastructure: creates the participant in the specified domain, sets up the AsyncWaitSet with
        // a configurable thread pool for event-driven processing, loads QoS profiles from the XML file,
        // and stores them for use by readers/writers
        auto participant_setup = std::make_shared<DDSParticipantSetup>(
            arguments.domain_id,
            ASYNC_WAITSET_THREADPOOL_SIZE,
            arguments.qos_file_path,
            qos_profiles::DEFAULT_PARTICIPANT,
            APP_NAME);
        
        // Setup DistLogger with the shared participant
        // DistLogger provides distributed logging over DDS network. By using the shared participant,
        // all log messages are published to remote subscribers via DDS topics, enabling centralized
        // logging and monitoring across distributed systems. This is more powerful than console logging.
        try {
            DistLoggerOptions options;
            options.domain_participant(participant_setup->participant());
            options.application_kind(APP_NAME);
            
            DistLogger::set_options(options);
            auto& dist_logger = DistLogger::get_instance();
            
            // Configure DistLogger verbosity and filter level for filtering which log messages to publish
            dist_logger.set_verbosity(rti::config::LogCategory::user, arguments.verbosity);
            dist_logger.set_filter_level(dist_logger.get_info_log_level());
            
            rti::config::Logger::instance().notice("DistLogger initialized with shared participant");
            rti::config::Logger::instance().notice(("Using QoS file: " + arguments.qos_file_path).c_str());
        } catch (const std::exception& ex) {
            std::cerr << "Error initializing DistLogger: " << ex.what() << std::endl;
            throw;
        }
        
        // Run the application
        run(participant_setup);
        
    } catch (const std::exception& ex) {
        // This will catch DDS exceptions
        std::cerr << "Exception in main: " << ex.what() << std::endl;
        return EXIT_FAILURE;
    }

    // Finalize Distributed Logger before participant factory
    try {
        rti::dist_logger::DistLogger::finalize();
        std::cout << "DistLogger finalized at application exit" << std::endl;
    } catch (const std::exception &e) {
        std::cerr << "Error finalizing DistLogger at exit: " << e.what() << std::endl;
    }

    // Finalize participant factory after all DDSParticipantSetup/DDSReaderSetup/DDSWriterSetup objects are destroyed
    // This should be called at application exit after all DDS entities are cleaned up
    try {
        dds::domain::DomainParticipant::finalize_participant_factory();
        std::cout << "DomainParticipant factory finalized at application exit" << std::endl;
    } catch (const std::exception &e) {
        std::cerr << "Error finalizing participant factory at exit: " << e.what() << std::endl;
    }

    return EXIT_SUCCESS;
}

File: apps/python/example_io_app/example_io_app.py
# (c) Copyright, Real-Time Innovations, 2025.  All rights reserved.
# RTI grants Licensee a license to use, modify, compile, and create derivative
# works of the software solely for use with RTI Connext DDS. Licensee may
# redistribute copies of the software provided that all such copies are subject
# to this license. The software is provided "as is", with no warranty of any
# type, including any warranty for fitness for any purpose. RTI is under no
# obligation to maintain or support the software. RTI shall not be liable for
# any incidental or consequential damages arising out of the use or inability
# to use the software.

import time
import sys
import os
import asyncio
import argparse
import rti.connextdds as dds
import rti.asyncio

# Add the DDS Python codegen path to Python path
# Prefer versioned types from DDS_PYTHON_GEN_DIR (set by run.sh)
_gen_dir = os.environ.get("DDS_PYTHON_GEN_DIR")
if _gen_dir and os.path.isdir(_gen_dir):
    sys.path.insert(0, _gen_dir)
else:
    # Fallback: use the checked-in types in dds/datamodel/
    sys.path.insert(
        0, os.path.join(os.path.dirname(__file__), "..", "..", "..", "dds", "datamodel")
    )

# Import DDS Data Types, Topics and config constants
from python_gen.ExampleTypes import example_types
from python_gen.Definitions import topics, qos_profiles, domains

# Application constants
PUBLISHER_SLEEP_INTERVAL = 2  # seconds for button/config publishing
# Published well above the downsampled_reader's 1Hz time-based filter so that
# app visibly demonstrates the downsampling (DataPatternsLibrary::Status1HzQoS).
POSITION_PUBLISH_INTERVAL = 0.1  # seconds (10Hz) for position publishing
MAIN_TASK_SLEEP_INTERVAL = 5  # seconds
DEFAULT_APP_NAME = "Example Python IO App"

async def process_command_data(reader):
    """Process incoming Command data"""
    # Print data as it arrives, suspending the coroutine until data is
    # available.
    async for data in reader.take_data_async():
        # Take all available command samples
        print(f"[COMMAND_SUBSCRIBER] Command Received:")
        print(f"  Command ID: {data.command_id}")
        print(f"  Destination ID: {data.destination_id}")
        print(f"  Command Type: {data.command_type}")
        print(f"  Message: {data.message}")
        print(f"  Urgent: {data.urgent}")


class ExampleIOApp:

    @staticmethod
    async def run(domain_id: int, qos_file_path: str):

        app_name = DEFAULT_APP_NAME

        # Load QoS profiles from the specified file
        print(f"Loading QoS profiles from: {qos_file_path}")
        qos_provider = dds.QosProvider(qos_file_path)

        # A DomainParticipant allows an application to begin communicating in
        # a DDS domain. Typically there is one DomainParticipant per application.
        # DomainParticipant QoS is configured in DDS_QOS_PROFILES.xml
        participant_qos = qos_provider.participant_qos_from_profile(
            qos_profiles.DEFAULT_PARTICIPANT
        )

        participant_qos.participant_name.name = app_name
        participant = dds.DomainParticipant(domain_id, participant_qos)

        print(
            f"DomainParticipant created with QoS profile: {qos_profiles.DEFAULT_PARTICIPANT}"
        )
        print(f"DOMAIN ID: {domain_id}")

        # Create Topics
        command_topic = dds.Topic(
            participant, topics.COMMAND_TOPIC, example_types.Command
        )

        button_topic = dds.Topic(participant, topics.BUTTON_TOPIC, example_types.Button)
        position_topic = dds.Topic(
            participant, topics.POSITION_TOPIC, example_types.Position
        )

        # Create DataReaders with QoS configured from DDS_QOS_PROFILES.xml
        # Using set_topic_data_reader_qos APUI allows us to use the external Assign QoS Profile
        # Otherwise can just use the regular datareader_qos_from_profile API
        command_reader_qos = qos_provider.set_topic_datareader_qos(
            qos_profiles.ASSIGNER, topics.COMMAND_TOPIC
        )
        command_reader = dds.DataReader(
            participant.implicit_subscriber, command_topic, command_reader_qos
        )

        # Create DataWriters with QoS configured from DDS_QOS_PROFILES.xml
        position_writer_qos = qos_provider.set_topic_datawriter_qos(
            qos_profiles.ASSIGNER, topics.POSITION_TOPIC
        )
        position_writer = dds.DataWriter(
            participant.implicit_publisher, position_topic, position_writer_qos
        )

        button_writer_qos = qos_provider.set_topic_datawriter_qos(
            qos_profiles.ASSIGNER, topics.BUTTON_TOPIC
        )
        button_writer = dds.DataWriter(
            participant.implicit_publisher, button_topic, button_writer_qos
        )

        print("[SUBSCRIBER] RTI Asyncio reader configured for Command data...")
        print(
            "[PUBLISHER] RTI Asyncio writers configured for Position and Button data..."
        )

        # Publisher coroutine for Position, published fast enough to
        # demonstrate the downsampled_reader's 1Hz time-based filter
        async def position_publisher_task():

            while True:
                try:
                    position_sample = example_types.Position()
                    position_sample.source_id = app_name
                    position_sample.latitude = 37.7749
                    position_sample.longitude = -122.4194
                    position_sample.altitude = 15.0
                    position_sample.timestamp_sec = int(time.time())

                    position_writer.write(position_sample)
                    print(
                        f"[POSITION_PUBLISHER] Published Position - Source: {position_sample.source_id}, Lat: {position_sample.latitude}, Lon: {position_sample.longitude}, Alt: {position_sample.altitude}"
                    )

                    await asyncio.sleep(POSITION_PUBLISH_INTERVAL)

                except asyncio.CancelledError:
                    print("[POSITION_PUBLISHER] Position publisher task cancelled")
                    break
                except Exception as e:
                    print(f"[POSITION_PUBLISHER] Error publishing data: {e}")
                    await asyncio.sleep(POSITION_PUBLISH_INTERVAL)

            print("[POSITION_PUBLISHER] Position publisher task finished.")

        # Publisher coroutine for Button, Config
        async def button_publisher_task():

            button_count = 0

            while True:
                try:
                    current_time = int(time.time())

                    # Publish Button message
                    button_sample = example_types.Button()
                    button_sample.source_id = app_name
                    button_sample.button_id = "btn_1"
                    button_sample.button_state = (
                        example_types.ButtonState.PRESSED
                    )
                    button_sample.press_count = button_count
                    button_sample.last_press_timestamp_sec = current_time
                    button_sample.hold_duration_sec = 0.0

                    button_writer.write(button_sample)
                    print(
                        f"[BUTTON_PUBLISHER] Published Button - ID: {button_sample.button_id}, Count: {button_count}"
                    )

                    button_count += 1

                    await asyncio.sleep(PUBLISHER_SLEEP_INTERVAL)

                except asyncio.CancelledError:
                    print("[BUTTON_PUBLISHER] Button publisher task cancelled")
                    break
                except Exception as e:
                    print(f"[BUTTON_PUBLISHER] Error publishing data: {e}")
                    await asyncio.sleep(PUBLISHER_SLEEP_INTERVAL)

            print("[BUTTON_PUBLISHER] Button publisher task finished.")

        # Main application coroutine
        async def main_task():
            count = 0
            while True:
                print(f"[MAIN] ExampleIOApp processing loop - iteration {count}")

                count += 1
                await asyncio.sleep(MAIN_TASK_SLEEP_INTERVAL)

        # Create and run concurrent tasks using RTI asyncio
        print("[MAIN] Starting RTI asyncio tasks...")
        try:
            await asyncio.gather(
                position_publisher_task(),
                button_publisher_task(),
                process_command_data(command_reader),
                main_task(),
            )
        except KeyboardInterrupt:
            print("[MAIN] Shutting down RTI asyncio tasks...")


def main():
    """Main entry point for the example_io_app application."""
    parser = argparse.ArgumentParser(
        description="Example I/O Application - Publishes Position/Button/Config, Subscribes to Command"
    )

    parser.add_argument(
        "-v",
        "--verbosity",
        type=int,
        default=1,
        help="Logging Verbosity",
    )

    parser.add_argument(
        "-d", "--domain_id", type=int, default=1, help="Domain ID (default: 1)"
    )

    parser.add_argument(
        "-q", "--qos_file", type=str, default="../../../dds/qos/DDS_QOS_PROFILES.xml",
        help="Path to QoS profiles XML file (default: ../../../dds/qos/DDS_QOS_PROFILES.xml)"
    )

    args = parser.parse_args()

    verbosity_levels = {
        0: dds.Verbosity.SILENT,
        1: dds.Verbosity.EXCEPTION,
        2: dds.Verbosity.WARNING,
        3: dds.Verbosity.STATUS_LOCAL,
        4: dds.Verbosity.STATUS_REMOTE,
        5: dds.Verbosity.STATUS_ALL,
    }

    # Sets verbosity for Connext Internals to help debugging
    verbosity = verbosity_levels.get(args.verbosity, dds.Verbosity.EXCEPTION)
    dds.Logger.instance.verbosity = verbosity

    try:
        # Run
        rti.asyncio.run(ExampleIOApp.run(domain_id=args.domain_id, qos_file_path=args.qos_file))
    except KeyboardInterrupt:
        pass


if __name__ == "__main__":
    main()
